src/core/transport/ssu_packet.cc

```diff
@@ -889,170 +861,130 @@ std::unique_ptr<SSUSessionDestroyedPacket> SSUPacketParser::ParseSessionDestroye
   return packet;
 }
 
-namespace SSUPacketBuilder {
+SSUPacketBuilder::SSUPacketBuilder(
+      std::uint8_t* data,
+      std::size_t len)
+  : OutputByteStream(data, len) {}
 
-void WriteData(
-    std::uint8_t*& pos,
-    const std::uint8_t* data,
-    std::size_t len) {
-  memcpy(pos, data, len);
-  pos += len;
-}
-
-void WriteUInt8(
-    std::uint8_t*& pos,
-    std::uint8_t data) {
-  *(pos++) = data;
-}
-
-void WriteUInt16(
-    std::uint8_t*& pos,
-    std::uint16_t data) {
-  htobe16buf(pos, data);
-  pos += 2;
-}
-
-void WriteUInt32(
-    std::uint8_t*& pos,
-    std::uint32_t data) {
-  htobe32buf(pos, data);
-  pos += 4;
-}
-
-std::size_t GetPaddingSize(
+std::size_t SSUPacketBuilder::GetPaddingSize(
     std::size_t size) {
   return (size % 16) ? 16 - size % 16 : 0;
 }
 
-std::size_t GetPaddedSize(
+std::size_t SSUPacketBuilder::GetPaddedSize(
     std::size_t size) {
   return size + GetPaddingSize(size);
 }
 
-void WriteHeader(
-    std::uint8_t*& data,
-    SSUHeader* header) {
+void SSUPacketBuilder::WriteHeader(SSUHeader* header) {
   if (header->GetMAC())
-    WriteData(data, header->GetMAC(), static_cast<std::size_t>(SSUSize::MAC));
+    WriteData(header->GetMAC(), static_cast<std::size_t>(SSUSize::MAC));
   else
-    data += static_cast<std::size_t>(SSUSize::MAC);  // Write MAC later
-  WriteData(data, header->GetIV(), static_cast<std::size_t>(SSUSize::IV));
+    ProduceData(static_cast<std::size_t>(SSUSize::MAC));  // Write MAC later
+  WriteData(header->GetIV(), static_cast<std::size_t>(SSUSize::IV));
   const std::uint8_t flag =
       (static_cast<std::uint8_t>(header->GetPayloadType()) << 4) +
       (header->HasRekey() << 3) +
       (header->HasExtendedOptions() << 2);
-  WriteUInt8(data, flag);
-  WriteUInt32(data, header->GetTime());
+  WriteUInt8(flag);
+  WriteUInt32(header->GetTime());
   if (header->HasExtendedOptions()) {
     // TODO(EinMByte): Check for overflow
-    WriteUInt8(data, header->GetExtendedOptionsSize());
+    WriteUInt8(header->GetExtendedOptionsSize());
     WriteData(
-        data,
         header->GetExtendedOptionsData(),
         header->GetExtendedOptionsSize());
   }
 }
 
-void WriteSessionRequest(
-    std::uint8_t*& buf,
-    SSUSessionRequestPacket* packet) {
-  WriteData(buf, packet->GetDhX(), static_cast<std::size_t>(SSUSize::DHPublic));
-  // TODO(EinMByte): Check for overflow
```

Comment:
`InputByteStream::ConsumeData(std::size_t amount)` and `OutputByteStream::ProduceData(std::size_t amount)` resolves these TODO's, yes?